parlai/core/torch_agent.py

```diff
@@ -1662,6 +1706,52 @@ def update_params(self):
         if self.opt.get('lr_scheduler') == 'invsqrt' and not self._is_lr_warming_up():
             # training step scheduler
             self.scheduler.step(self._number_training_updates)
+        elif self.opt.get('lr_scheduler') == 'cosine' and not self._is_lr_warming_up():
+            # The cosine scheduler is a funny beast. It has the LR multiplier go
+            # from 1 .. 0 on a cosine half period. Then it jumps immediately back
+            # to 1 and goes down to 0 again. There are two additional
+            # modifications. On the end of the period, we do two things: we
+            # multiply the length of the period by --lr-period-mult and we
+            # multiply the maximum learning rate by --lr-decay-rate. You can
+            # see an illustration of --lr-period-mult here:
+            # https://cdn-images-1.medium.com/max/1600/1*dMvU26iOxbkQ9sTqhaWrsg.png
+
+            if not hasattr(self, '_lr_period'):
+                raise RuntimeError('Looks like you forgot to call build_lr_scheduler')
+
+            # how many steps since warmup?
+            warmup_updates = max(0, self.opt['warmup_updates'])
+            cosine_steps = self._number_training_updates - warmup_updates
+
+            # factor we lengthen the cosine period by
+            lr_mult = self.opt['lr_period_mult']
+
+            # passes contains the number of periods we have iterated
+            # step contains the number of steps relative to the last reset
+            if lr_mult == 1:
+                # special case, do this
+                passes = cosine_steps // self._lr_period
+                step = cosine_steps % self._lr_period
+            else:
+                # total number of cycles completed
+                # see http://mathworld.wolfram.com/ExponentialSumFormulas.html
+                n = cosine_steps / self._lr_period
+                passes = np.floor(np.log((lr_mult - 1) * n + 1) / np.log(lr_mult))
+                last_reset_step = (
+                    self._lr_period * (1 - lr_mult ** passes) / (1 - lr_mult)
+                )
+                step = cosine_steps - last_reset_step
+                # cycle_steps contains the total length of this period
+                cycle_steps = self._lr_period * lr_mult ** passes
+                self.scheduler.T_max = cycle_steps
+
+            # the height of the peak of the cosine wave is multiplied by
+            # decay every period
+            mult_max = self.opt['lr_scheduler_decay'] ** passes
+            lr = self.opt['learningrate']
+            self.scheduler.base_lrs = [lr * mult_max for _ in self.scheduler.base_lrs]
```

Comment:
what does this do?

Reply:
It's the way you set the peak learning rate. Every time you iterate through a cosine (half period), the coefficient resets to 1, but the max rate decays. I should scare up an illustration.

Reply:
Okay here's how it looks. This corresponds to

`--lr-scheduler-decay 0.75 --lr-period-mult 2 --lr-period-updates 1000`

![Screen Shot 2019-07-10 at 10 51 03 AM](https://user-images.githubusercontent.com/31896/60979392-a67c4400-a300-11e9-8b66-b2946151f52d.png)

Notice at the end of a period, the peak is multiplied by `0.75` and the length of the new period is doubled.

Reply:
nice thanks for the picture

Reply:
The paper sometimes uses a schedule/not a multiplicative decay factor (i.e. 0.050, 0.025, 0.01 and 0.005) - do we care?

Reply:
I don’t think so. I’m mimicking fairseq for this bit. Both come out to roughly the same. With decay of 0.5 you get (.5, .25, .125. .0625, ...) so it’s the right ballpark

Reply:
Sure - more a question of if there are times when it wouldn't be similar. Say you might want something other than multiplicative (i.e. a defined schedule or log or something like that). I'm happy to leave it though if you are. I need to get calibrated as to what extent we care about such things!